src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumneManager.java

```diff
@@ -0,0 +1,173 @@
+  private final SnowflakeServiceClient snowflakeServiceClient;
+
+  // Client prefix generated by the Snowflake server
+  private final String clientPrefix;
+
+  /**
+   * Constructor for ExternalVolumeManager
+   *
+   * @param isTestMode whether the manager in test mode
+   * @param role the role of the client
+   * @param clientName the name of the client
+   * @param snowflakeServiceClient the Snowflake service client used for configure calls
+   */
+  ExternalVolumeManager(
+      boolean isTestMode,
+      String role,
+      String clientName,
+      SnowflakeServiceClient snowflakeServiceClient) {
+    this.role = role;
+    this.clientName = clientName;
+    this.snowflakeServiceClient = snowflakeServiceClient;
+    this.externalVolumeMap = new ConcurrentHashMap<>();
+    try {
+      this.clientPrefix =
+          isTestMode
+              ? "testPrefix"
+              : this.snowflakeServiceClient
+                  .clientConfigure(new ClientConfigureRequest(role))
+                  .getClientPrefix();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.CLIENT_CONFIGURE_FAILURE, e.getMessage());
+    }
+  }
+
+  /**
+   * Given a fully qualified table name, return the target storage by looking up the table name
+   *
+   * @param fullyQualifiedTableName the target fully qualified table name
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage<T, ExternalVolumeLocation> getStorage(
+      String fullyQualifiedTableName) {
+    // Only one chunk per blob in Iceberg mode.
+    StreamingIngestStorage<T, ExternalVolumeLocation> stage =
+        this.externalVolumeMap.get(fullyQualifiedTableName);
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
+          String.format("No external volume found for table %s", fullyQualifiedTableName));
+    }
+
+    return stage;
+  }
+
+  /**
+   * Add a storage to the manager by looking up the table name from the open channel response
+   *
+   * @param dbName the database name
+   * @param schemaName the schema name
+   * @param tableName the table name
+   * @param fileLocationInfo response from open channel
+   */
+  @Override
+  public void addStorage(
+      String dbName, String schemaName, String tableName, FileLocationInfo fileLocationInfo) {
+    String fullyQualifiedTableName =
+        Utils.getFullyQualifiedTableName(dbName, schemaName, tableName);
+
+    try {
+      this.externalVolumeMap.put(
+          fullyQualifiedTableName,
+          new StreamingIngestStorage<T, ExternalVolumeLocation>(
+              this,
+              this.clientName,
+              fileLocationInfo,
+              new ExternalVolumeLocation(dbName, schemaName, tableName),
+              DEFAULT_MAX_UPLOAD_RETRIES));
+    } catch (SnowflakeSQLException | IOException err) {
+      throw new SFException(err, ErrorCode.UNABLE_TO_CONNECT_TO_STAGE);
```

Comment:
Could we add table name to the error message?

Reply:
Added tableName in the error message.